pkg/controller/virtualmachineimport/virtualmachineimport_controller.go

```diff
@@ -358,6 +370,84 @@ func (r *ReconcileVirtualMachineImport) upsertStatusConditions(vmiName types.Nam
 	return nil
 }
 
+func (r *ReconcileVirtualMachineImport) storeSourceVMStatus(instance *v2vv1alpha1.VirtualMachineImport, vmStatus string) error {
+	vmiCopy := instance.DeepCopy()
+	vmiCopy.Annotations[sourceVMInitialState] = vmStatus
+
+	patch := client.MergeFrom(instance)
+	return r.client.Patch(context.TODO(), vmiCopy, patch)
+}
+
+//TODO: use in proper places
+func (r *ReconcileVirtualMachineImport) afterFailure(vmiName types.NamespacedName, p provider.Provider) error {
+	var errs []error
+	err := r.afterImport(p, vmiName.Namespace)
+	if err != nil {
+		errs = append(errs, err)
+	}
+	err = r.restoreInitialVMState(vmiName, p)
+	if err != nil {
+		errs = append(errs, err)
+	}
+
+	if len(errs) > 0 {
+		message := ""
+		for _, e := range errs {
+			message = utils.WithMessage(message, e.Error())
+		}
+		return fmt.Errorf("Import failure clean-up for %v failed: %s", vmiName, message)
+	}
+	return nil
+}
+
+func (r *ReconcileVirtualMachineImport) restoreInitialVMState(vmiName types.NamespacedName, p provider.Provider) error {
+	var instance v2vv1alpha1.VirtualMachineImport
+	err := r.client.Get(context.TODO(), vmiName, &instance)
+	if err != nil {
+		return err
+	}
+	if vmInitialState, ok := instance.Annotations[sourceVMInitialState]; ok {
+		if vmInitialState == string(provider.VMStatusUp) {
```

Comment:
you can reduce one nesting level of if if you return quickly in case `ok` is false

Reply:
Good idea. Fixed.